### ISSUE TITLE
bootcamp app 2.0 migration test: PDP does not display much after the upgrade 

### ISSUE THREAD

**Issue by plabadie** [SAP/spartacus#7621]
After the 2.0 upgrade, the PDP does not display much.  See screenshot.

To repro just click on only product from the home page.

I do not see any errors that give clue to what can be happening.

![Screen Shot 2020-05-13 at 5 40 42 PM](https://user-images.githubusercontent.com/7065346/81869289-c80e1380-9541-11ea-8f42-4273596fb743.png)

To repro:

I used dev17 as a backend

I used the config below in app.module.ts to enable most of the bootcamp features.

I followed the steps in https://github.com/SAP/spartacus-docs/blob/v2-develop/_pages/home/updating-to-version-2/updating-to-version-2.md

```

@NgModule({
  declarations: [AppComponent],
  imports: [
    BrowserModule.withServerTransition({ appId: 'spartacus-app' }),
    B2cStorefrontModule.withConfig({
      backend: {
        occ: {
          baseUrl: environment.occBaseUrl,
        },
      },
      context: {
        baseSite: ['electronics-spa'],
      },
    }),

    // demonstrate i18n
    ConfigModule.withConfig({
      i18n: {
        resources: translations,
        chunks: translationChunksConfig,
        fallbackLang: 'en',
      },
    }),

    CustomBannerModule,

    // BootcampWebComponentModule,

    BootcampRoutingModule,

    // // Module to configure the site context with multi-site features
    // BootcampMultiSiteModule,
    // // Demonstrate a custom context
    CatalogContextModule,

    BootcampSeoModule,

    BootcampStateModule,

    // BootcampDataBindingModule,

    // Demonstrate static CMS config
    BootcampStaticCmsModule,

    BootcampConfigModule,
    // // Provide a custom them to the `ThemeConfig`.
    ConfigModule.withConfig({ theme: Theme.STRAWBERRIES } as ThemeConfig),

    // // Demonstrate outlets. Don't forget to enable the `<app-outlets></app-outlets>` in the app.component.html
    BootcampOutletModule,

    BootcampIconModule,

    BootcampI18nModule,

    // demonstrates layout configuration
    BootcampLayoutConfigModule,

    BootcampComponentModule,
  ],
  providers: [],
  bootstrap: [AppComponent],
})
export class AppModule {}
```

**Comment by znikola** [comment]
For me it works with both dev17 and ccv2 instance. 
<img width="611" alt="Screen Shot 2020-05-15 at 12 23 55" src="https://user-images.githubusercontent.com/11561723/82040503-f822f700-96a6-11ea-8312-28d99018bb7a.png">

However, I did find one issue related to the ssr migration which is fixed by: https://github.com/SAP/spartacus/pull/7655 . In case this is not merged by the time you want to test it again, I suggest to checkout this PR's branch and publish it to _verdaccio_. 
In case you'd like to test it locally (without using the public npm registry) you can try referring to the updated schematics README, which is hopefully more clear than the previous version: https://github.com/SAP/spartacus/pull/7601

I did all the steps described in the issue for that PR (https://github.com/SAP/spartacus/issues/7640), and additionally I did the following:
- in the `src/app/features/seo/custom-content-page-meta.resolver.ts` I replaced the contents of `resolve(): Observable<PageMeta>` with just `of()`.

**Comment by znikola** [comment]
I believe this is caused when we include all the modules in the app shell.
I don't think it's an update bug, but we can maybe investigate it further. What do you think @plabadie ?

**Comment by plabadie** [comment]
The issue is in the `BootcampRoutingModule`, more specifically this exact PRODUCT_NORMALIZER config:

https://github.com/SAP/spartacus-bootcamp/blob/master/src/app/features/routing/product-routes.module.ts#L52-L56

When I comment it out, the PDP works.  This worked with 1.x, it stopped working after the 2.0 upgrade.  I don't know what the normalizer does that isn't working well with 2.0.  Needs investigation to determine if this breaking change is covered in our docs.

**Comment by znikola** [comment]
It seems that this caused by routing when we want to use a custom property as part of the route: https://github.com/SAP/spartacus-bootcamp/blob/05eaf36d3bb8fd031042673f2ad6fc6109a8b34d/src/app/features/routing/product-category.normalizer.ts#L23

@Platonn do you have any idea what's causing this?
Maybe it's a breaking change that we are not aware of.

**Comment by Platonn** [comment]
**TLDR;**
You just need to wrap the logic of the `ProductPrettyNameNormalizer` in `if(source.name)`:
```typescript
export class ProductPrettyNameNormalizer
  implements Converter<Occ.Product, Product> {
  convert(source: Occ.Product, target?: any): Product {
    if (source.name) {
      target.prettyName = source.name.replace(/ /g, '-');
    }
    return target;
  }
}
```
It would be good to update the `master` branch in the bootcamp repo, since it can be used in future I guess.

--- 
Since the product loading scopes were introduced (see docs https://sap.github.io/spartacus-docs/loading-scopes/), the `PRODUCT_NORMALIZER`s computations are execuded on each product entity, no matter how much stripped out of data it is.
In particular the Product Classifications scope contains only the property `classifications` (but not product code or name).
That's why the above normalizer throws an error and it breaks the final RxJs stream of data.

Open for discussion:
1. Should we document that customers have to be careful when accessing `source` data? IMO definitely yes
1. Should we precise the NORMALZIERS to run only _for a specific loading scopes_? (architectural change). IMO rather no
1. Should we *handle* RxJs errors for each NORMALIZER not to break the entire stream on error? IMO rather yes, but can be postponed

**Comment by znikola** [comment]
After some investigation, I found out that the error is "swallowed" in the `projects/core/src/product/store/effects/product.effect.ts`'s `loadProduct$` effect. We don't log the error nor we put it in the state, so there's no feedback of what went wrong.
This is a good topic candidate for 3.0.
